exporter dies with panic: label value [...] is not valid UTF-8 when having interface comments with Umlauts

When scraping a Mikrotik v7.2.3, the exporter dies when an Interface comment contains an Umlaut (German ü) with the following error message: 

panic: label value "078xxxxxxx 50GB/Month 12.50Fr. , Monat beginnt jeweils am 9. Abo bis 16.08.2023, 60 Tage im Voraus k\xfcnden." is not valid UTF-8

goroutine 35 [running]:
github.com/prometheus/client_golang/prometheus.MustNewConstMetric(...)
        /root/go/pkg/mod/github.com/prometheus/client_golang@v1.4.1/prometheus/value.go:105
mikrotik-exporter/collector.(*interfaceCollector).collectMetricForProperty(0x7fae8e5d3008?, {0x889f93, 0xa}, 0xc0000daf80, 0xc00000e270)
        /var/lib/prometheus/mikrotik-exporter/mikrotik-exporter/collector/interface_collector.go:104 +0x71d
mikrotik-exporter/collector.(*interfaceCollector).collectForStat(...)
        /var/lib/prometheus/mikrotik-exporter/mikrotik-exporter/collector/interface_collector.go:66
mikrotik-exporter/collector.(*interfaceCollector).collect(0xc0000ca500, 0xc000198180?)
        /var/lib/prometheus/mikrotik-exporter/mikrotik-exporter/collector/interface_collector.go:45 +0xfa
mikrotik-exporter/collector.(*collector).connectAndCollect(0xc0000da9c0, 0xc000198180, 0xc000070240)
        /var/lib/prometheus/mikrotik-exporter/mikrotik-exporter/collector/collector.go:355 +0x36d
mikrotik-exporter/collector.(*collector).collectForDevice(0xc000222768?, {{0x7fffb6995840, 0x9}, {0x7fffb6995819, 0xc}, {{0x0, 0x0}, {{0x0, 0x0}, 0x0}}, ...}, ...)
        /var/lib/prometheus/mikrotik-exporter/mikrotik-exporter/collector/collector.go:326 +0xe5
mikrotik-exporter/collector.(*collector).Collect.func1({{0x7fffb6995840, 0x9}, {0x7fffb6995819, 0xc}, {{0x0, 0x0}, {{0x0, 0x0}, 0x0}}, {0x7fffb6995850, ...}, ...})
        /var/lib/prometheus/mikrotik-exporter/mikrotik-exporter/collector/collector.go:291 +0x78
created by mikrotik-exporter/collector.(*collector).Collect
        /var/lib/prometheus/mikrotik-exporter/mikrotik-exporter/collector/collector.go:290 +0x285

Unfortunately, my golang skills are very limited, I would have tried to fix it myself if I could..

Here's what I was able to figure out so far:
- The prometheus client_golang@v1.4.1 accepts UTF-8 only in prometheus/value.go
- Therefore, mikrotik-exporter ( interface_collector.go ) should convert the string in label "comment" to UTF-8 first (and maybe additionally emit a warning that the string was modified)

--- a/collector/helper.go
+++ b/collector/helper.go
@@ -7,6 +7,7 @@ import (
 	"strconv"
 	"strings"
 	"time"
+	"unicode/utf8"
 
 	"github.com/prometheus/client_golang/prometheus"
 	log "github.com/sirupsen/logrus"
@@ -22,6 +23,17 @@ func init() {
 
 func metricStringCleanup(in string) string {
 	return strings.Replace(in, "-", "_", -1)
+}
+
+// cleanLabelValue replaces invalid UTF-8 byte sequences with the Unicode
+// replacement character so that free-form RouterOS fields such as interface
+// comments, which may use a non-UTF-8 encoding, do not crash
+// prometheus.MustNewConstMetric.
+func cleanLabelValue(in string) string {
+	if utf8.ValidString(in) {
+		return in
+	}
+	return strings.ToValidUTF8(in, "�")
 }
 
 func descriptionForPropertyName(prefix, property string, labelNames []string) *prometheus.Desc {

--- a/collector/helper_test.go
+++ b/collector/helper_test.go
@@ -3,9 +3,27 @@ package collector
 import (
 	"math"
 	"testing"
+	"unicode/utf8"
 
+	"github.com/prometheus/client_golang/prometheus"
 	"github.com/stretchr/testify/assert"
 )
+
+func TestCleanLabelValue(t *testing.T) {
+	// "künden." encoded as Windows-1252/Latin-1, as returned by some RouterOS devices.
+	invalid := "k\xfcnden."
+
+	assert.False(t, utf8.ValidString(invalid))
+
+	cleaned := cleanLabelValue(invalid)
+	assert.True(t, utf8.ValidString(cleaned))
+	assert.Equal(t, "künden.", cleanLabelValue("künden."))
+
+	desc := prometheus.NewDesc("test_metric", "", []string{"comment"}, nil)
+	assert.NotPanics(t, func() {
+		prometheus.MustNewConstMetric(desc, prometheus.GaugeValue, 1, cleaned)
+	})
+}
 
 func TestSplitStringToFloats(t *testing.T) {
 	var testCases = []struct {

--- a/collector/interface_collector.go
+++ b/collector/interface_collector.go
@@ -102,7 +102,8 @@ func (c *interfaceCollector) collectMetricForProperty(property string, re *proto
 			}
 		}
 		ctx.ch <- prometheus.MustNewConstMetric(desc, vtype, v, ctx.device.Name, ctx.device.Address,
-			re.Map["name"], re.Map["type"], re.Map["disabled"], re.Map["comment"], re.Map["running"], re.Map["slave"])
+			cleanLabelValue(re.Map["name"]), cleanLabelValue(re.Map["type"]), cleanLabelValue(re.Map["disabled"]),
+			cleanLabelValue(re.Map["comment"]), cleanLabelValue(re.Map["running"]), cleanLabelValue(re.Map["slave"]))
 
 	}
 }

